### DISCUSSION TITLE
A quick update on the status of development.

### DISCUSSION THREAD

**Discussion by Animeshz** [Animeshz/EgisTec-EH575#1] [General]
Unfortunately, the init and repeat sequence of the sensor does not seem to be a constant like the other models so far which had some tries of reverse engineering.

The init sequence sent/received in the Windows most of the time is:
```
Packet    7 out 45:47:49:53:60:00:fc
Packet   10  in 53:49:47:45:00:aa:01
Packet   11 out 45:47:49:53:60:01:fc
Packet   14  in 53:49:47:45:01:00:01
Packet   15 out 45:47:49:53:60:40:fc
Packet   18  in 53:49:47:45:40:00:01
Packet   19 out 45:47:49:53:63:09:0b:83:24:00:44:0f:08:20:20:01:05:12
Packet   22  in 53:49:47:45:09:0b:01:83:24:00:44:0f:08:20:20:01:05:12
Packet   23 out 45:47:49:53:63:26:06:06:60:06:05:2f:06
Packet   26  in 53:49:47:45:26:06:01:06:60:06:05:2f:06
Packet   27 out 45:47:49:53:61:23:00
Packet   30  in 53:49:47:45:23:00:01
Packet   31 out 45:47:49:53:61:24:33
Packet   34  in 53:49:47:45:24:33:01
Packet   35 out 45:47:49:53:61:20:00
Packet   38  in 53:49:47:45:20:00:01
Packet   39 out 45:47:49:53:61:21:66
Packet   42  in 53:49:47:45:21:66:01
Packet   43 out 45:47:49:53:60:00:66
Packet   46  in 53:49:47:45:00:aa:01
Packet   47 out 45:47:49:53:60:01:66
Packet   50  in 53:49:47:45:01:00:01
Packet   51 out 45:47:49:53:63:2c:02:00:57
Packet   54  in 53:49:47:45:2c:02:01:00:57
Packet   55 out 45:47:49:53:60:2d:02
Packet   58  in 53:49:47:45:2d:47:01
Packet   59 out 45:47:49:53:62:67:03
Packet   62  in 53:49:47:45:67:03:01:7d:87:81
Packet   63 out 45:47:49:53:60:0f:03
Packet   66  in 53:49:47:45:0f:20:01
Packet   67 out 45:47:49:53:63:2c:02:00:13
Packet   70  in 53:49:47:45:2c:02:01:00:13
Packet   71 out 45:47:49:53:60:00:02
Packet   74  in 53:49:47:45:00:ab:01
Packet   75 out 45:47:49:53:64:14:ec
Packet   78  in <image data>
```

Whereas most of time in the linux, there are some deviations, as follows:
```
Packet 14 in: [53, 49, 47, 45, 1, 1, 1]
Packet 18 in: [53, 49, 47, 45, 40, 80, 1]
Packet 50 in: [53, 49, 47, 45, 1, 1, 1]
Packet 62 in: [53, 49, 47, 45, 67, 3, 1, 0, 0, 0]
```

And hence the returned image is full of zeros (complete black).

I'm currently getting into the disassembly to study the original driver's logic over the input sequences to find out correct output sequence, I gathered few information, but a lot of information is under function pointers contained under struct/classes which has lost its static type information and hence is hard to navigate and find function with known signature. Some function calls are done by incrementing pointers, which neither gives information about itself nor about the function it calls.

I'll try to compensate the delay by making a working PoC as soon as possible<sup>1</sup> with minimum customization options and simple algorithm without much enhancements on raw image sent by sensor, which may not work very well with slightly less-pressed finger and low-quality captures, and will push updates later.

<sub>1. I also have few entrance examinations coming up in few weeks, so will add less hours on this, so can't guarantee a ETA as of now, but I'll try to cover this up to the best extent as soon as I get more time.</sub>

**Comment by AnkushSinghGandhi** [comment]
hey i want to ask about progress

**Reply by Animeshz** [reply]
@AnkushSinghGandhi I've collected all the required sequences (I guess) by running windows vm under vmware in different linux hosts & also on windows host under different conditions. The image captures are working fine, found USB wakeup switching sequence as well using that. Reversing the original driver did not much helped in obtaining those sequences as static memory did not had sequences at one place and navigating forward through virtual functions was somewhat easy but navigating back (finding which function called current function) was difficult (even though I connected more than 60% of the functions to their classes using ghidra, some python script and pharos/ooanalyzer, but it was taking forever to obtain small bit of information).

I had like 3 exams back to back each with about 1 week gap 16Jul, 25Jul, and next is tomorrow 8th Aug, so I did not got much time in the implementation, after tomorrow I'd get a break of about 3 weeks so can work on... I see the 1c7a:0570 driver just got merged into libfprint so I'd guess I'll just do the same for now, even if [#271](https://gitlab.freedesktop.org/libfprint/libfprint/-/issues/271) and [#272](https://gitlab.freedesktop.org/libfprint/libfprint/-/issues/272) are still open, the [#170 for Egis 570](https://gitlab.freedesktop.org/libfprint/libfprint/-/merge_requests/170) got merged and driver was able to perform well.

This repository may soon be archived with all the efforts in finding the relevant information (scripts, data, logs, and will document few things I've found about the driver while reversing like the registry keys).

**Reply by dev0x13** [reply]
Hi Animesh. First of all, thank you very much for your work! Secondly - good luck with your tomorrow exam!
For some time I have been trying to adopt the 1c7a:0570 dirver from libfprint for Egis 575 using the packet sequences you've provided above, but I did not manage to make image capturing work. Initialization seems to work fine though. I don't want to distract you from the preparation for exam and from your break, but may I ask you to share the packet sequences you've obtained if it won't require much efforts please? Thank you in advance! I would greatly appreciate this!

**Reply by Animeshz** [reply]
@dev0x13

```
45:47:49:53:60:00:00
45:47:49:53:60:01:00
45:47:49:53:61:0a:fd
45:47:49:53:61:35:02
45:47:49:53:61:80:00
45:47:49:53:60:80:00
45:47:49:53:61:0a:fc
45:47:49:53:63:01:02:0f:03
45:47:49:53:61:0c:22
45:47:49:53:61:09:83
45:47:49:53:63:26:06:06:60:06:05:2f:06
45:47:49:53:61:0a:f4
45:47:49:53:61:0c:44
45:47:49:53:61:50:03
45:47:49:53:60:50:03
45:47:49:53:73:14:ec
45:47:49:53:60:40:ec
45:47:49:53:63:09:0b:83:24:00:44:0f:08:20:20:01:05:12
45:47:49:53:63:26:06:06:60:06:05:2f:06
45:47:49:53:61:23:00
45:47:49:53:61:24:33
45:47:49:53:61:20:00
45:47:49:53:61:21:66
45:47:49:53:60:00:66
45:47:49:53:60:01:66
45:47:49:53:60:40:66
45:47:49:53:61:0c:22
45:47:49:53:61:0b:03
45:47:49:53:61:0a:fc
45:47:49:53:60:00:fc  # start here if
45:47:49:53:60:01:fc
45:47:49:53:60:40:fc  # response to this comes 40:00:01 or else start from top
45:47:49:53:63:09:0b:83:24:00:44:0f:08:20:20:01:05:12
45:47:49:53:63:26:06:06:60:06:05:2f:06
45:47:49:53:61:23:00
45:47:49:53:61:24:33
45:47:49:53:61:20:00
45:47:49:53:61:21:66
45:47:49:53:60:00:66
45:47:49:53:60:01:66
45:47:49:53:63:2c:02:00:57
45:47:49:53:60:2d:02
45:47:49:53:62:67:03
45:47:49:53:60:0f:03
45:47:49:53:63:2c:02:00:13
45:47:49:53:60:00:02
45:47:49:53:64:14:ec
```

This sequence will work, adding a jupyter analysis notebook with all possible data I collected (see readme in logs dir)
[wireshark_analysis_1.zip](https://github.com/Animeshz/EgisTec-EH575/files/6948961/wireshark_analysis_1.zip)

**Edit:** few samples recently taken using the burst_capture.rs in the repo with slight modification...
![image](https://user-images.githubusercontent.com/26714676/128597790-9f1c1e55-7abd-4d4c-a997-320e7757fcd2.png)

**Reply by dev0x13** [reply]
Thank you very much!
